File: browser/src/js/seller/stack.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

const WebsocketLocation = require(
    '../moneysocket/beacon/location/websocket.js').WebsocketLocation;
const OutgoingWebsocketLayer = require(
    "../moneysocket/protocol/websocket/outgoing_layer.js")
        .OutgoingWebsocketLayer;
const ProviderLayer = require(
    "../moneysocket/protocol/provider/layer.js").ProviderLayer;
const OutgoingRendezvousLayer = require(
    "../moneysocket/protocol/rendezvous/outgoing_layer.js")
        .OutgoingRendezvousLayer;
const ProviderTransactLayer = require(
    "../moneysocket/protocol/transact/provider_layer.js").ProviderTransactLayer;
const SellerLayer = require("../seller/layer.js").SellerLayer;


class SellerStack {
    constructor(app, ui) {
        this.app = app;
        this.ui = ui;

        this.onnexusonline = null;
        this.onnexusoffline = null;
        this.onstackevent = null;
        this.handleinvoicerequest = null;
        this.handlepayrequest = null;
        this.handleproviderinforequest = null;
        this.handleopinioninvoicerequest = null;
        this.handlesellerinforequest = null;

        this.websocket_layer = this.setupOutgoingWebsocketLayer();
        this.rendezvous_layer = this.setupOutgoingRendezvousLayer(
            this.websocket_layer);
        this.provider_layer = this.setupProviderLayer(this.rendezvous_layer);
        this.transact_layer = this.setupProviderTransactLayer(
            this.provider_layer);
        this.seller_layer = this.setupSellerLayer(this.transact_layer);

        this.seller_layer.onnexusonline = (function(nexus) {
            this.announceNexus(nexus);
        }).bind(this);
        this.seller_layer.onnexusoffline = (function(nexus) {
            this.revokeNexus(nexus);
        }).bind(this);

        this.nexus = null;
        this.shared_seed = null;
    }

    //////////////////////////////////////////////////////////////////////////
    // setup
    //////////////////////////////////////////////////////////////////////////

    setupSellerLayer(below_layer) {
        var l = new SellerLayer();
        l.onlayerevent = (function(nexus, status) {
            this.onLayerEvent("SELLER", nexus, status);
        }).bind(this);
        l.handleopinioninvoicerequest = (function(nexus, msats, request_uuid) {
           this.handleOpinionInvoiceRequest(msats, request_uuid);
        }).bind(this);
        l.handlesellerinforequest = (function() {
           return this.handleSellerInfoRequest();
        }).bind(this);
        l.registerAboveLayer(below_layer);
        return l;
    }

    setupProviderTransactLayer(below_layer) {
        var l = new ProviderTransactLayer();
        l.onlayerevent = (function(nexus, status) {
            this.onLayerEvent("PROVIDER_TRANSACT", nexus, status);
        }).bind(this);
        l.handleinvoicerequest = (function(nexus, msats, request_uuid) {
            this.handleInvoiceRequest(msats, request_uuid);
        }).bind(this);
        l.handlepayrequest = (function(nexus, msats, request_uuid) {
            this.handlePayRequest(msats, request_uuid);
        }).bind(this);
        l.registerAboveLayer(below_layer);
        return l;
    }

    setupProviderLayer(below_layer) {
        var l = new ProviderLayer();
        l.onlayerevent = (function(nexus, status) {
            this.onLayerEvent("PROVIDER", nexus, status);
        }).bind(this);
        l.handleproviderinforequest = (function(shared_seed) {
            return this.handleProviderInfoRequest(shared_seed);
        }).bind(this);
        l.registerAboveLayer(below_layer);
        return l;
    }

    setupOutgoingRendezvousLayer(below_layer) {
        var l = new OutgoingRendezvousLayer();
        l.onlayerevent = (function(nexus, status) {
            this.onLayerEvent("OUTGOING_RENDEZVOUS", nexus, status);
        }).bind(this);
        l.registerAboveLayer(below_layer);
        return l;
    }

    setupOutgoingWebsocketLayer() {
        var l = new OutgoingWebsocketLayer();
        l.onlayerevent = (function(nexus, status) {
            this.onLayerEvent("OUTGOING_WEBSOCKET", nexus, status);
        }).bind(this);
        return l;
    }

    //////////////////////////////////////////////////////////////////////////

    handleProviderInfoRequest(shared_seed) {
        console.assert(this.handleproviderinforequest != null);
        return this.handleproviderinforequest();
    }

    providerNowReadyFromApp() {
        this.provider_layer.providerNowReadyFromApp();
    }

    sendProviderInfoUpdate() {
        if (this.nexus == null) {
            return;
        }
        var shared_seed = this.nexus.getSharedSeed();
        this.provider_layer.sendProviderInfoUpdate(shared_seed);
    }

    sellerNowReadyFromApp() {
        this.seller_layer.sellerNowReadyFromApp();
    }

    updatePrices() {
        this.nexus.updatePrices();
    }

    //////////////////////////////////////////////////////////////////////////
    // layer callbacks:
    //////////////////////////////////////////////////////////////////////////

    announceNexus(below_nexus) {
        console.log("provider stack got nexus");
        this.nexus = below_nexus;
        this.shared_seed = below_nexus.getSharedSeed();

        if (this.onnnexusonline != null) {
            this.onnexusonline(below_nexus);
        }
    }

    revokeNexus(below_nexus) {
        console.log("provider stack got nexus revoked");
        this.nexus = null;
        this.shared_seed = null;
        if (this.onnnexusoffline != null) {
            this.onnexusoffline(below_nexus);
        }
    }

    onLayerEvent(layer_name, nexus, status) {
        if (this.onstackevent != null) {
            this.onstackevent(layer_name, nexus, status);
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // seller layer callbacks
    ///////////////////////////////////////////////////////////////////////////

    handleOpinionInvoiceRequest(item_id, request_uuid) {
        this.handleopinioninvoicerequest(item_id, request_uuid);
    }

    fulfilOpinionInvoiceRequest(bolt11, request_reference_uuid) {
        this.nexus.notifyOpinionInvoice(bolt11, request_reference_uuid);
    }

    handleSellerInfoRequest() {
        return this.handlesellerinforequest();
    }

    fulfilOpinion(item_id, opinion, request_reference_uuid) {
        this.nexus.notifyOpinion(item_id, opinion, request_reference_uuid);
    }

    ///////////////////////////////////////////////////////////////////////////
    // transact layer allbacks
    ///////////////////////////////////////////////////////////////////////////

    handleInvoiceRequest(msats, request_uuid) {
        if (this.handleinvoicerequest != null) {
            this.handleinvoicerequest(mstats, request_uuid);
        }
    }

    fulfilRequestInvoice(bolt11, request_reference_uuid) {
        this.nexus.notifyInvoice(bolt11, request_reference_uuid);
    }

    handlePayRequest(bolt11, request_uuid) {
        if (this.handlepayrequest != null) {
            this.handlepayrequest(bolt11, request_uuid);
        }
    }

    fulfilRequestPay(preimage, request_reference_uuid) {
        this.nexus.notifyPreimage(preimage, request_reference_uuid);
    }

    //////////////////////////////////////////////////////////////////////////
    // UI calls these:
    //////////////////////////////////////////////////////////////////////////

    doConnect(beacon) {
        console.log("provider layer connect called");

        var location = beacon.locations[0];
        var shared_seed = beacon.getSharedSeed();
        if (! (location instanceof WebsocketLocation)) {
            console.error("not websocket location?");
            return;
        }
        console.log("connect provider: " + location.toWsUrl());
        this.websocket_layer.connect(location, shared_seed);
    }

    doDisconnect() {
        console.log("provider layer disconnect called");
        this.websocket_layer.initiateCloseAll();
    }
}

exports.SellerStack = SellerStack;

File: python/terminus/stack.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

from moneysocket.layer.provider import ProviderLayer
from moneysocket.layer.rendezvous.outgoing import OutgoingRendezvousLayer
from moneysocket.layer.websocket.outgoing import OutgoingWebsocketLayer
from moneysocket.layer.local.outgoing import OutgoingLocalLayer

from moneysocket.stack.incoming import IncomingStack

from terminus.layer import TerminusLayer

class TerminusStack(object):
    def __init__(self, config):
        self.config = config

        self.onannounce = None
        self.onrevoke = None
        self.onstackevent = None
        self.handleproviderinforequest = None
        self.handleinvoicerequest = None
        self.handlepayrequest = None

        # two ways to make a connection: 1) via incoming websocket (that comes
        # through the incoming stack and pass via the local layer) and
        # 2) via outgoing websocket connections we initiate.
        self.local_layer = self.setup_local_layer()
        self.websocket_layer = self.setup_websocket_layer()
        self.rendezvous_layer = self.setup_rendezvous_layer(
            self.websocket_layer, self.local_layer)
        self.provider_layer = self.setup_provider_layer(self.rendezvous_layer)
        self.terminus_layer = self.setup_terminus_layer(self.provider_layer)
        self.incoming_stack = self.setup_incoming_stack(self.config,
                                                        self.local_layer)

    ###########################################################################

    def setup_terminus_layer(self, below_layer):
        l = TerminusLayer()
        l.register_above_layer(below_layer)
        l.register_layer_event(self.send_stack_event, "TERMINUS")
        l.handleinvoicerequest = self.terminus_handle_invoice_request
        l.handlepayrequest = self.terminus_handle_pay_request
        l.handleproviderinforequest = self.terminus_handle_provider_info_request
        return l

    def setup_provider_layer(self, below_layer):
        l = ProviderLayer()
        l.register_above_layer(below_layer)
        l.register_layer_event(self.send_stack_event, "PROVIDER")
        l.handleproviderinforequest = self.terminus_handle_provider_info_request
        return l

    def setup_rendezvous_layer(self, below_layer_1, below_layer_2):
        l = OutgoingRendezvousLayer()
        l.register_above_layer(below_layer_1)
        l.register_above_layer(below_layer_2)
        l.register_layer_event(self.send_stack_event, "OUTGOING_RENDEZVOUS")
        return l

    def setup_websocket_layer(self):
        l = OutgoingWebsocketLayer()
        l.register_layer_event(self.send_stack_event, "OUTGOING_WEBSOCKET")
        return l

    def setup_local_layer(self):
        l = OutgoingLocalLayer()
        l.register_layer_event(self.send_stack_event, "OUTGOING_LOCAL")
        return l

    def setup_incoming_stack(self, config, local_layer):
        s = IncomingStack(config, local_layer)
        return s

    ###########################################################################

    def announce_nexus(self, terminus_nexus):
        if self.onannounce:
            self.onannounce(terminus_nexus)

    def revoke_nexus(self, terminus_nexus):
        if self.onrevoke:
            self.onrevoke(terminus_nexus)

    def send_stack_event(self, layer_name, nexus, status):
        if self.onstackevent:
            self.onstackevent(layer_name, nexus, status)

    ###########################################################################

    def notify_preimage(self, shared_seeds, preimage):
        self.terminus_layer.notify_preimage(shared_seeds, preimage)

    ###########################################################################

    def terminus_handle_provider_info_request(self, shared_seed):
        assert self.handleproviderinforequest
        return self.handleproviderinforequest(shared_seed)

    def terminus_handle_pay_request(self, shared_seed, bolt11):
        assert self.handlepayrequest
        self.handlepayrequest(shared_seed, bolt11)

    def terminus_handle_invoice_request(self, shared_seed, msats):
        assert self.handleinvoicerequest
        return self.handleinvoicerequest(shared_seed, msats)

    ###########################################################################

    def connect(self, location, shared_seed):
        return self.websocket_layer.connect(location, shared_seed)

    def disconnect(self, shared_seed):
        self.websocket_layer.disconnect(shared_seed)

    ###########################################################################

    def local_connect(self, shared_seed):
        self.local_layer.connect(shared_seed)

    def local_disconnect(self, shared_seed):
        self.local_layer.disconnect(shared_seed)

    ###########################################################################

    def listen(self):
        self.incoming_stack.listen()

    def get_listen_locations(self):
        return self.incoming_stack.get_listen_locations()
